File: Solutions/src/Solution1.java
// Time Complexity :
//KlogK
//// Space Complexity :O(n)

// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : None
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

class Solution1 {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> results = new ArrayList<>();
        HashMap<String, List<String>> map = new HashMap<>();
        Arrays.sort(strs);
        for (int i=0;i<strs.length;i++){
            String temp = strs[i];
            char[] ch = temp.toCharArray();
            Arrays.sort(ch);
            if (map.containsKey(String.valueOf(ch))){
                map.get(String.valueOf(ch)).add(strs[i]);
            }
            else{
                List<String> tmpres = new ArrayList<>();
                tmpres.add(strs[i]);
                map.put(String.valueOf(ch), tmpres);
            }

        }

        for (List<String> item: map.values()) {
            results.add(item);
        }
        return results;
    }
}

File: solution1.py
# Time Complexity : O(NK) where N is length of strs list and K is max length of a string in strs
# Space Complexity : O(NK)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach


import collections

class Solution:
    def group_anagrams(self, strs):
        
        ans=collections.defaultdict(list)
        
        for s in strs:
            count=[0]*26
            for c in s:
                count[ord(c)-ord('a')]+=1
            # tuple is hashable
            ans[tuple(count)].append(s)
        return ans.values()